src/EventTeachingService.cpp

```diff
@@ -11,259 +11,427 @@
+            index = module_config->findEventSpace();
+          }
+
+          // if existing or new event space found, write the event data
+          if (index < module_config->EE_MAX_EVENTS) 
+          {
+            // write the event to EEPROM at this location -- EVs are indexed from 1 but storage offsets start at zero !!
+            // DEBUG_SERIAL << F("> writing EV = ") << evindex << F(", at index = ") << index << F(", offset = ") << (module_config->EE_EVENTS_START + (index * module_config->EE_BYTES_PER_EVENT)) << endl;
+
+            // don't repeat this for subsequent EVs
+            if (evindex < 2) 
+            {
+              module_config->writeEvent(index, &msg->data[1]);
+            }
+
+            module_config->writeEventEV(index, evindex, evval);
+
+            // recreate event hash table entry
+            // DEBUG_SERIAL << F("> updating hash table entry for idx = ") << index << endl;
+            module_config->updateEvHashEntry(index);
+
+            // respond with WRACK
+            controller->sendWRACK();  // Deprecated in favour of GRSP_OK
+            controller->sendGRSP(OPC_EVLRN, getServiceID(), GRSP_OK);
+          } 
+          else 
+          {
+            // DEBUG_SERIAL << F("> no free event storage, index = ") << index << endl;
+            // respond with CMDERR & GRSP
+            controller->sendCMDERR(CMDERR_TOO_MANY_EVENTS);
+            controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_TOO_MANY_EVENTS);
+          }
+        }
       }
 
-      // if existing or new event space found, write the event data
-      if (index < module_config->EE_MAX_EVENTS)
-      {
-        // write the event to EEPROM at this location -- EVs are indexed from 1 but storage offsets start at zero !!
-        // DEBUG_SERIAL << F("> writing EV = ") << evindex << F(", at index = ") << index << F(", offset = ") << (module_config->EE_EVENTS_START + (index * module_config->EE_BYTES_PER_EVENT)) << endl;
+      return PROCESSED;
 
-        // don't repeat this for subsequent EVs
-        if (evindex < 2)
+    case OPC_EVLRNI:
+      // F5 - learn an event using event index
+
+      DEBUG_SERIAL << endl
+                   << F("> EVLRNI for source nn = ") << nn << endl;
+
+      // we must be in learn mode
+      if (bLearn) 
+      {
+        if (msg->len < 8) 
+        {
+          controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_INV_CMD);
+          controller->sendCMDERR(CMDERR_INV_CMD);
+        } 
+        else 
         {
-          module_config->writeEvent(index, &msg->data[1]);
+          byte index = msg->data[5];
+          byte evIndex = msg->data[6];
+          byte evVal = msg->data[7];
+
+          // invalid index
+          if (index >= module_config->EE_MAX_EVENTS) 
+          {
+            DEBUG_SERIAL << F("> invalid index") << endl;
+            controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_INV_EV_IDX);
+            return PROCESSED;
```

Comment:
This return is superflous as you have a return at the end of the case-block. 
You don't have a return in the other branches. I don't care which style you use, just want it to be consistent. :)

Reply:
Good point.